### ISSUE TITLE
COVIDSafe can't access JWT token on startup (crash)

### ISSUE THREAD

**Issue by jimmo** [AU-COVIDSafe/mobile-android#23]
Affects COVIDSafe v1.0.49

This has happened to three friends of mine, and today I was finally able to get a stack trace. See stack traces below, but it fails in getEncrypterJWTToken while trying to do the AEAD decryption using the key from keystore and the encrypted JWT token from SharedPreferences.

I think this might be related to installing updates (i.e. something about the update causing the app to be stopped) as in all three cases it coincided with a new version being released, and the result appears to be that either:
 - The keystore is corrupted and the AES256_GCM_SPEC key gets lost (and silently recreated when accessed)
 - The SharedPreferences are corrupted and the (encrypyed) JWT token cannot be accessed.

The second case seems surprising because
 - SharedPreferences are committed atomically (even though the code uses apply(), writes to SharedPreferences should be committed atomically).
 - I not sure that the JWT is the first thing that's accessed?
 - The JWT is only ever written once, so why would just the JWT be corrupted due to a lost write?

The keystore also seems surprising because nothing else uses that key but the fact that it silently recreates the key would explain why decryption failed.

Anyway, no idea. But this must be happening frequently because pretty much everyone I know who has the app has now seen this. This is very serious because there's no way to recover other than to clear all app data which of course loses the encounter log history and requires a re-registration. Futhermore, if this is related to the app being stopped by the updater, then this is going to be a bigger issue with the new push notifications (and whatever you're planning to do about getting older versions updated).

This appears to be the same issue as https://twitter.com/_the_culture/status/1281505842026577920?s=21

I also reported this privately by email on 29 July but no response. It's disappointing that a fix wasn't put in place in 1.0.39 or 1.0.48 to at least allow-reregistration to get a new JWT while keeping the encounter log.

Two different stack traces showing the same issue are below.

```
08-03 12:51:03.299 29604 29604 E AndroidRuntime: FATAL EXCEPTION: main
08-03 12:51:03.299 29604 29604 E AndroidRuntime: Process: au.gov.health.covidsafe, PID: 29604
08-03 12:51:03.299 29604 29604 E AndroidRuntime: java.lang.SecurityException: Could not decrypt value. decryption failed
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.security.crypto.EncryptedSharedPreferences.getDecryptedObject(EncryptedSharedPreferences.java:546)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.security.crypto.EncryptedSharedPreferences.getString(EncryptedSharedPreferences.java:377)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.Preference.getEncrypterJWTToken(Preference.kt:100)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.interactor.usecase.UpdateBroadcastMessageAndPerformScanWithExponentialBackOff.run(UpdateBroadcastMessageAndPerformScanWithExponentialBackOff.kt:25)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.interactor.usecase.UpdateBroadcastMessageAndPerformScanWithExponentialBackOff.run(UpdateBroadcastMessageAndPerformScanWithExponentialBackOff.kt:17)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.interactor.UseCase$invoke$1$result$1.invokeSuspend(UseCase.kt:26)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at kotlinx.coroutines.scheduling.CoroutineScheduler.runSafely(CoroutineScheduler.kt:571)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.executeTask(CoroutineScheduler.kt:738)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.runWorker(CoroutineScheduler.kt:678)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at kotlinx.coroutines.scheduling.CoroutineScheduler$Worker.run(CoroutineScheduler.kt:665)
08-03 12:51:03.299 29604 29604 E AndroidRuntime: Caused by: java.security.GeneralSecurityException: decryption failed
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at com.google.crypto.tink.aead.AeadWrapper$WrappedAead.decrypt(AeadWrapper.java:83)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    at au.gov.health.covidsafe.security.crypto.EncryptedSharedPreferences.getDecryptedObject(EncryptedSharedPreferences.java:499)
08-03 12:51:03.299 29604 29604 E AndroidRuntime:    ... 11 more
```

```
08-07 12:01:50.363 22560 22560 E AndroidRuntime: FATAL EXCEPTION: main
08-07 12:01:50.363 22560 22560 E AndroidRuntime: Process: au.gov.health.covidsafe, PID: 22560
08-07 12:01:50.363 22560 22560 E AndroidRuntime: java.lang.SecurityException: Could not decrypt value. decryption failed
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at au.gov.health.covidsafe.security.crypto.EncryptedSharedPreferences.getDecryptedObject(EncryptedSharedPreferences.java:546)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at au.gov.health.covidsafe.security.crypto.EncryptedSharedPreferences.getString(EncryptedSharedPreferences.java:377)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at au.gov.health.covidsafe.Preference.getEncrypterJWTToken(Preference.kt:101)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at au.gov.health.covidsafe.scheduler.GetMessagesScheduler.getMessages(GetMessagesScheduler.kt:102)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at au.gov.health.covidsafe.scheduler.GetMessagesJobSchedulerService$onStartJob$1$1.invokeSuspend(GetMessagesScheduler.kt:60)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at android.os.Handler.handleCallback(Handler.java:883)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at android.os.Handler.dispatchMessage(Handler.java:100)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at android.os.Looper.loop(Looper.java:214)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at android.app.ActivityThread.main(ActivityThread.java:7356)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at java.lang.reflect.Method.invoke(Native Method)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:492)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:930)
08-07 12:01:50.363 22560 22560 E AndroidRuntime: Caused by: java.security.GeneralSecurityException: decryption failed
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at com.google.crypto.tink.aead.AeadWrapper$WrappedAead.decrypt(AeadWrapper.java:83)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    at au.gov.health.covidsafe.security.crypto.EncryptedSharedPreferences.getDecryptedObject(EncryptedSharedPreferences.java:499)
08-07 12:01:50.363 22560 22560 E AndroidRuntime:    ... 13 more
```

**Comment by gavintapp2** [comment]
Issue created with stack trace on github https://github.com/AU-COVIDSafe/mobile-android/issues/23

**Comment by chrisculnane** [comment]
@jimmo and myself have been looking into this further, and it looks to be a known issue with the AndroidKeyStore that is not gracefully handled by Tink, see: https://github.com/google/tink/issues/321

An update to Tink has been released to more gracefully handle the underlying issue with the AndroidKeyStore, see the release notes for 1.4.0: https://github.com/google/tink/releases/tag/v1.4.0. (COVIDSafe is currently referencing 1.4.0-rc2). However, note the following extract from the release notes:

> Refactoring Android Keystore integration
> - Running a self-test to only enable the integration if Android Keystore is working properly.
> - Do not automatically generate fresh keys if keys exist but are corrupt. This won't recover corrupt keys, but at least it will allow to gather more data on Android Keystore failures.

There is a related androidx.security issue: https://issuetracker.google.com/issues/158234058, which may be relevant given that EncryptedSharedPreferences is copied from that library. There's an updated release of androidx.security that references the new version of Tink, as well as containing an unrelated update to EncryptedSharedPreferences, see https://developer.android.com/jetpack/androidx/releases/security#1.1.0-alpha02, which may want to be included in COVIDSafe's copy of EncryptedSharedPreferences.

**Comment by thaidn** [comment]
Yeah this is a known issue in Tink versions earlier than 1.4.0. Upgrading to 1.4.0 is recommended. Lost keys can't be recovered though.

**Comment by chrisculnane** [comment]
@thaidn after some further digging and discussions with @jimmo we're not sure just updating Tink will be sufficient.

We were wondering why the reports in https://issuetracker.google.com/issues/158234058 indicated that the update to AndroidX Security 1.0.0 RC 2 seemed to trigger more occurrences of the issue. That update adds support for MasterKey and deprecates the `create` method in EncryptedSharedPreferences that uses the masterKeyAlias in favour of the one that accepts a MasterKey. MasterKeys and MasterKey both directly access the AndroidKeyStore without using Tink. As such, the tests introduced in Tink to catch problems with accessing the AndroidKeyStore won't help in those two classes, and as such, won't prevent the creation of new master keys when there is an problem with accessing the AndroidKeyStore.

From a COVIDSafe point of view, is the call to MasterKeys `val masterKeyAlias = MasterKeys.getOrCreate(MasterKeys.AES256_GCM_SPEC)` in Preference.kt even necessary? Wouldn't Tink's AndroidKeySetManager perform the same actions but with added checks for issues accessing the AndroidKeyStore? It seems it might be better to not use MasterKeys and only access through Tink to benefit from the additional checks in 1.4.0.

COVIDSafe will also probably need to update Preference.kt to handle the occasional failure to access the AndroidKeyStore and apply a random back-off and retry.

**Comment by jimmo** [comment]
It's unclear whether this explains why it appears to be related to app updates, but potentially what's happening is that every on TempID fetch (i.e. call to getEncrypterJTWToken) it has a chance to see the transient keystore error and therefore replace the key. If this happens while the app is backgrounded then it will crash without the user noticing. This means that auto-update can now happen as the app is not running. The user then notices when they first re-open the app and see the crash.

Since 1.0.48, the TempId now happens 24x less often (daily) so the rate of incidence of this issue might fall?

**Comment by covidsafe-support** [comment]
Hi @jimmo and @chrisculnane - thanks for creating this issue and the ensuing leads, we have been working on addressing this and appreciate the information you are able to provide.

**Comment by covidsafe-team** [comment]
Hi @jimmo  and @chrisculnane - thanks for your patience. This was fixed in release 1.0.54
